src/levenberg_marquardt.jl

```diff
@@ -98,52 +98,33 @@ function levenberg_marquardt(df::OnceDifferentiable, initial_x::AbstractVector{T
         # jacobian! will check if x is new or not, so it is only actually
         # evaluated if x was updated last iteration.
         jacobian!(df, x) # has alias J
-
-        # we want to solve:
+        # Use QR with column pivoting to solve the regularized least squares problem
         #    argmin 0.5*||J(x)*delta_x + f(x)||^2 + lambda*||diagm(J'*J)*delta_x||^2
-        # Solving for the minimum gives:
-        #    (J'*J + lambda*diagm(DtD)) * delta_x == -J' * f(x), where DtD = sum(abs2, J,1)
-        # Where we have used the equivalence: diagm(J'*J) = diagm(sum(abs2, J,1))
-        # It is additionally useful to bound the elements of DtD below to help
-        # prevent "parameter evaporation".
-
-        DtD = vec(sum(abs2, J, dims=1))
-        for i in 1:length(DtD)
-            if DtD[i] <= MIN_DIAGONAL
-                DtD[i] = MIN_DIAGONAL
-            end
-        end
-
-        # delta_x = ( J'*J + lambda * Diagonal(DtD) ) \ ( -J'*value(df) )
-        mul!(JJ, transpose(J), J)
-        @simd for i in 1:n
-            @inbounds JJ[i, i] += lambda * DtD[i]
+        Q,R,p = qr(J, Val(true))
+        rhs = -Matrix(Q)'*value(df)
+        if isreal(R)
```

Comment:
`isreal` is not what you want here. Instead check on the scalar type (eg `eltype(R) <: Complex`)

Reply:
Maybe, ```eltype(J) <: Complex``` is better yet